### DISCUSSION TITLE
<title>Let me know how to use GPU when running YOLOv7 on CVAT

### DISCUSSION THREAD

**Discussion by tlswhd0215** [community/community#103129] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Body

annotation is working but gpu is not used

first time using cvat and docker 

# nvidia-smi 
Fri Feb  2 13:26:09 2024       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.40.06              Driver Version: 551.23         CUDA Version: N/A      |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 2070 ...    On  |   00000000:01:00.0  On |                  N/A |
|  0%   48C    P0             41W /  215W |     956MiB /   8192MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|  No running processes found                                                             |
+-----------------------------------------------------------------------------------------+

nuctl deploy result

24.02.02 22:00:31.300                     nuctl (I) Deploying function {"name": "onnx-wongkinyiu-yolov7"}
24.02.02 22:00:31.304                     nuctl (I) Building {"builderKind": "docker", "versionInfo": "Label: 1.11.24, Git commit: f2a3900d23b92fd3639dc9cb765044ef53a4fb2b, OS: linux, Arch: amd64, Go version: go1.19.10", "name": "onnx-wongkinyiu-yolov7"}
24.02.02 22:00:31.496                     nuctl (I) Staging files and preparing base images
24.02.02 22:00:31.508                     nuctl (I) Building processor image {"registryURL": "", "taggedImageName": "cvat/onnx-wongkinyiu-yolov7:latest"}
24.02.02 22:00:31.508     nuctl.platform.docker (I) Pulling image {"imageName": "quay.io/nuclio/handler-builder-python-onbuild:1.11.24-amd64"}
24.02.02 22:00:35.596     nuctl.platform.docker (I) Pulling image {"imageName": "quay.io/nuclio/uhttpc:0.0.1-amd64"}
24.02.02 22:00:39.949            nuctl.platform (I) Building docker image {"image": "cvat/onnx-wongkinyiu-yolov7:latest"}
24.02.02 22:00:40.484            nuctl.platform (I) Pushing docker image into registry {"image": "cvat/onnx-wongkinyiu-yolov7:latest", "registry": ""}
24.02.02 22:00:40.484            nuctl.platform (I) Docker image was successfully built and pushed into docker registry {"image": "cvat/onnx-wongkinyiu-yolov7:latest"}
24.02.02 22:00:40.484                     nuctl (I) Build complete {"image": "cvat/onnx-wongkinyiu-yolov7:latest"}
24.02.02 22:00:40.490                     nuctl (I) Cleaning up before deployment {"functionName": "onnx-wongkinyiu-yolov7"}
24.02.02 22:00:41.679            nuctl.platform (I) Waiting for function to be ready {"timeout": 120}
24.02.02 22:00:43.086                     nuctl (I) Function deploy complete {"functionName": "onnx-wongkinyiu-yolov7", "httpPort": 33000, "internalInvocationURLs": ["172.18.0.20:8080"], "externalInvocationURLs": ["0.0.0.0:33000"]}

https://github.com/opencv/cvat/blob/develop/serverless/onnx/WongKinYiu/yolov7/nuclio/function-gpu.yaml

**Comment by gil-son** [comment]
here are a few things to check:
- GPU Support in Docker
- GPU Usage in the Function
- Function Code
- Function Logs
- Update Docker Images

**Comment by chiarorosa** [comment]
@tlswhd0215 Deploy the function using Nuclio CLI, making sure that the GPU resources are correctly allocated. Run the following command: 
`nuctl deploy --project-name cvat --path serverless/onnx/WongKinYiu/yolov7 --platform local --resource-limit nvidia.com/gpu=1`

you need to ensure that your Docker containers are correctly configured to access the GPU.
- Ensure NVIDIA Container Toolkit is Installed
- Modify Docker Compose File

something like that
`version: '3.7'
services:
  cvat:
    image: cvat/server
    runtime: nvidia
    environment:
      - NVIDIA_VISIBLE_DEVICES=all
      - NVIDIA_DRIVER_CAPABILITIES=compute,utility
    ...`
